### DISCUSSION TITLE
Adapting the model to Polish

### DISCUSSION THREAD

**Discussion by KarolinaPondel** [speechbrain/speechbrain#2618] [Q&A]
Hello,
I am a PhD student from Poland, with a background in electronic acoustics, I am working on my PhD in ASR. I would like to use SpeechBrain toolkit in my thesis, my work is in the field of acoustics and I use neural networks as a tool to study signals. I initially went through the templates/speech_recognition/ASR example and by following the instructions I managed to train a model for the Polish language dataset, for 100 epochs I obtained a WER of 30%. In your repository I found pre-trained ASR transformer models based on CommonVoice for German, Italian and French. I have been trying to adapt this model to the Polish version of CommonVoice, unfortunately for many weeks without success. In the common_voice_prepare.py script I added code to decode Polish characters, modified the yaml file specifying as data_folder the path to my data, experimented with different output neuron sizes and batch_size. I observe that when increasing the number of epochs, the model does not learn, the loss drops, increases, drops etc, and the model decodes the same syllable for all samples. Unfortunately I don't have a person in my area who specialises in this type of model and could help me with this problem. I am learning everything myself... I have also not found tutorials on how to train this particular model. Perhaps I am making a mistake that I am unable to verify myself. I would very much appreciate your help, contact, consultation.
Best regards,
Karolina PS

**Comment by asumagic** [comment]
Hello, the CommonVoice transformer recipes might be a decent place to start indeed. Recently, (streaming) conformers for CommonVoice have also been added, with okay results.

Have you tried using subword tokenization (since reading this this reads like you are using characters)? Have you maybe changed defaults on batch size or multi-GPU training? This can have an effect on transformer model training (e.g. warmup becomes too short or batch size is too low if gradient accumulation was not compensated for).

Is there some "obvious" pattern of failure when you check the resulting WER text files, which show per-utterance errors? (e.g. some character is always missed, tokens don't look as they should, the last word is always wrong)

**Reply by asumagic** [reply]
Ah, sorry, I somehow missed "and the model decodes the same syllable for all samples". This is relatively strange, can you send the training script and hyperparams to see if there's anything weird?

**Reply by KarolinaPondel** [reply]
As for the type of tokenization - I experimented with different types (“unigram”, “bpe”, “char”), the result was always the same. I changed batch_size to 12, due to the performance of the graphics cards I use (3x RTX A6000, 48 GB GPU each). I run the learning as: torchrun --nproc_per_node=3 train.py train.yaml (as explained here https://colab.research.google.com/drive/13pBUacPiotw1IvyffvGZ-HrtBr9T6l15?usp=sharing). “Obvious” pattern of failure when checking the resulting WER text files: each word is decoded as the same syllable, this syllable changed with the next test, but the test trend the same.

I'm sending my scripts:
train.py: https://gist.github.com/KarolinaPondel/1374252a1411500e4f534625b12e012d
train.yaml: https://gist.github.com/KarolinaPondel/6c7a48ca2d58e7f2424b4e0f73cf6bef
common_voice_prepare.py, where I added information about Polish characters: https://gist.github.com/KarolinaPondel/9f144a70ac807f85e5b1c10431d90d4f
fragment of file wer_test.txt after 100 epoch with output_neurons: 1000, The original file has 45066 lines, for each recording the result is the same: https://gist.github.com/KarolinaPondel/c4e1c914f1af2c083cc434cb0c10fc26

**Reply by KarolinaPondel** [reply]
@asumagic In the meantime I ran the script provided in the speechbrain repository for French, I used CommonVoice 18.0 for this language, 100 epochs, batch size = 32. Unfortunately the results are meaningless, for each recording the model returns the same 2 characters... At 54 epochs, the train loss stopped falling and remained constant until the end of learning.

**Reply by asumagic** [reply]
Sorry for the delay, I'm currently out of office. It is strange that the CV recipe is failing to converge completely even for French, since I think this model got trained again recently (at least CTC and transducer ones).

In general, you should be seeing the WER go <90% within less than 10 epochs. (And usually it doesn't go >100% WER from hallucination either at really any point)

The batch size seems reasonable.

Which SpeechBrain version do you have installed?

cc @TParcollet I think you've touched those models recently?

**Reply by KarolinaPondel** [reply]
I installed SpeechBrain on the begining of April this year by cloning github repo, my WER after 100 epoch is 100%. I am trying to train: speechbrain/recipes/CommonVoice/ASR/transformer with train_fr.yaml and train.py provided in your repo.
